primitives/staking/src/lib.rs

```diff
@@ -69,10 +75,55 @@ pub struct Stake<T: StakingInterface + ?Sized> {
 	/// This is only guaranteed to reflect the amount locked by the staking system. If there are
 	/// non-staking locks on the bonded pair's balance this amount is going to be larger in
 	/// reality.
-	pub total: T::Balance,
+	pub total: Balance,
 	/// The total amount of the stash's balance that will be at stake in any forthcoming
 	/// rounds.
-	pub active: T::Balance,
+	pub active: Balance,
+}
+
+/// A generic staking event listener. Currently used for implementations involved in stake tracking.
+/// Note that the interface is designed in a way that the events are fired post-action, so any
+/// pre-action data that is needed needs to be passed to interface methods.
+/// The rest of the data can be retrieved by using `StakingInterface`.
+pub trait OnStakingUpdate<AccountId, Balance> {
+	/// Track ledger updates.
+	fn on_update_ledger(who: &AccountId, prev_stake: Option<Stake<AccountId, Balance>>);
+	/// Track nominators, those reinstated and also new ones.
+	fn on_nominator_add(who: &AccountId, prev_nominations: Vec<AccountId>);
+	/// Track validators, those reinstated and new.
+	fn on_validator_add(who: &AccountId);
+	/// Track removed validators. Either chilled or those that became nominators instead.
+	fn on_validator_remove(who: &AccountId); // only fire this event when this is an actual Validator
+	/// Track removed nominators.
+	fn on_nominator_remove(who: &AccountId, nominations: Vec<AccountId>); // only fire this if this is an actual Nominator
+	/// Track those participants of staking system that are kicked out for whatever reason.
+	fn on_reaped(who: &AccountId); // -> basically `kill_stash`
+}
+
+impl<AccountId, Balance> OnStakingUpdate<AccountId, Balance> for () {
```

Comment:
```suggestion
#[cfg(feature = "std")]
impl<AccountId, Balance> OnStakingUpdate<AccountId, Balance> for () {
```

to make sure it is not used by accident in the real runtime.